packages/punchcard/lib/appsync/vtl-object.ts

```diff
@@ -0,0 +1,208 @@
+import { AnyShape, BinaryShape, bool, integer, NeverShape, NothingShape, ShapeGuards, ShapeVisitor, UnknownShape, Value } from '@punchcard/shape';
+import { ArrayShape, BoolShape, DynamicShape, IntegerShape, MapShape, NumberShape, Pointer, RecordShape, SetShape, Shape, StringShape, TimestampShape } from '@punchcard/shape';
+import { string, Trait } from '@punchcard/shape';
+import { FunctionArgs, FunctionShape } from '@punchcard/shape/lib/function';
+import { VExpression } from './expression';
+
+const type = Symbol.for('GraphQL.Type');
+const expr = Symbol.for('GraphQL.Expression');
+
+export class VObject<T extends Shape = Shape> {
+  public static readonly Type = type;
+  public static readonly Expr = expr;
+
+  public static typeOf<T extends VObject>(t: T): T[typeof type] {
+    return t[type];
+  }
+
+  public static exprOf<T extends VObject>(t: T): T[typeof expr] {
+    return t[expr];
+  }
+
+  public readonly [type]: T;
+  public readonly [expr]: VExpression;
+  constructor(_type: T, _expr: VExpression) {
+    this[type] = _type;
+    this[expr] = _expr;
+  }
+
+  public toJson(): VString {
+    return new VString(string, new VExpression(ctx => `$util.toJson(${VObject.exprOf(this).visit(ctx)})`));
+  }
+}
+
+export namespace VObject {
+  export type TypeOf<T extends VObject> = T[typeof type];
+
+  export function isObject(a: any): a is VObject {
+    return a[expr] !== undefined;
+  }
+
+  // export type ShapeOf<T extends VObject> = T extends VObject<infer I> ? I : never;
+
+  export type Of<T extends Shape> =
+    T extends RecordShape<infer M> ? VRecord<{
+      [m in keyof M]: Of<M[m]>;
+    }> & {
+      [m in keyof M]: Of<M[m]>;
+    } :
+    T extends ArrayShape<infer I> ? VList<VObject.Of<I>> :
+    T extends MapShape<any> ? never : // maps are not supported in GraphQL
+    T extends BoolShape ? VBool :
+    T extends DynamicShape<any> ? VAny :
+    T extends IntegerShape ? VInteger :
+    T extends NumberShape ? VFloat :
+    T extends StringShape ? VString :
+    T extends TimestampShape ? VTimestamp :
+
+    VObject<T>
+    ;
+
+  /**
+   * Object that is "like" a VObject for some Shape.
+   *
+   * Like meaning that is either an expression, or a collection
+   * of expressions that share the structure of the target type.
+   */
+  export type Like<T extends Shape> = Value.Of<T> | VObject.Of<T> | (
+    T extends RecordShape<infer M> ? {
+      [m in keyof M]: Like<Pointer.Resolve<M[m]>>;
+    } :
+    T extends ArrayShape<infer I> ? Like<I>[] :
+    T extends SetShape<infer I> ? Like<I>[] :
+    T extends MapShape<infer I> ? {
+      [key: string]: Like<I>;
+    } :
+    VObject.Of<T> | Value.Of<T>
+  );
+}
+
+export const IDTrait: {
+  [Trait.Data]: {
+    graphqlType: 'ID'
+  }
+} = {
+  [Trait.Data]: {
+    graphqlType: 'ID'
+  }
+};
+
+export const ID = string.apply(IDTrait);
```

Comment:
When extending this for other GraphQL use cases, it'll be important to enable non-string IDs. What other approaches could you take to creating the ID shape?